### PR TITLE
Use ElementTree in glifLib and drop embedded XML parsing infrastructure

### LINKED ISSUE #16: [glifLib] raise warn("...", DeprecationWarning) ?
In `glifLib` module, there are improper usages of `warnings.warn` function.

```
$ git grep "raise warn("
Lib/ufoLib/glifLib.py:1142:             raise warn("The beginPath method needs an identifier kwarg. The contour's identifier value has been discarded.", DeprecationWarning)
Lib/ufoLib/glifLib.py:1166:                     raise warn("The addPoint method needs an identifier kwarg. The point's identifier value has been discarded.", DeprecationWarning)
Lib/ufoLib/glifLib.py:1196:             raise warn("The addComponent method needs an identifier kwarg. The component's identifier value has been discarded.", DeprecationWarning)
```

`warnings.warn` function doesn't have a return value, it simply returns `None`. Therefore, trying to call `raise warn(...)` will issue a `TypeError: exceptions must derive from BaseException`, as if one attempts to do `raise None`.

https://docs.python.org/2/library/warnings.html#warnings.warn
https://docs.python.org/3.5/library/warnings.html#warnings.warn

You must either raise a proper exception, or you can use warnings filters to turn matching warnings into exceptions: https://docs.python.org/3.5/library/warnings.html#the-warnings-filter

### LINKED ISSUE #17: formatVersions is not used in readGlyphFromString
The `formatVersions` argument is in the `readGlyphFromString` function's signature as well as in the docstring. However it is not used inside the function's body. I just wonder why.

https://github.com/unified-font-object/ufoLib/blob/master/Lib/ufoLib/glifLib.py#L499

I suggest the following change:

```diff
diff --git a/Lib/ufoLib/glifLib.py b/Lib/ufoLib/glifLib.py
index a87a9f0..eb00080 100755
--- a/Lib/ufoLib/glifLib.py
+++ b/Lib/ufoLib/glifLib.py
@@ -496,7 +496,7 @@ def readGlyphFromString(aString, glyphObject=None, pointPen=None, formatVersions
 	that are allowed to be read.
 	"""
 	tree = _glifTreeFromFile(BytesIO(tobytes(aString)))
-	_readGlyphFromTree(tree, glyphObject, pointPen)
+	_readGlyphFromTree(tree, glyphObject, pointPen, formatVersions=formatVersions)
 
 
 def writeGlyphToString(glyphName, glyphObject=None, drawPointsFunc=None, writer=None, formatVersion=2):

```

### PR DESCRIPTION
**Pull request by adrientetar**
- I think there is further optimization path with plistlib. It's a bit too tied to xmlTreeBuilder and could perhaps use a C implementation.
- I guess we could drop the fontTools XML writer when writing (we don't need a custom XML reader/writer these days?)

There are some assertRaises tests that fail even though the assertions are raised in the code (I tried following the code paths...) I didn't change anything around testing so I'm not sure what's causing this. 
cc @moyogo @anthrotype 

Closes #16.
Closes #17.
cc #27 (should the issue be salvaged for further optimizations?)

### CODE REVIEW

**Comment by adrientetar** [comment]
The startup time for TruFont with a ~700glyphs font went from 1.45s to 1.23s with this PR.

**Comment by adrientetar** [comment]
If I disable the reading of GLIF's plist, it went from 1.23s to 0.76s! I'm sure plist parsing can be improved.

**Comment by typesupply** [comment]
Wow, thanks! I'll review this in the next few days.

> I guess we could drop the fontTools XML writer when writing (we don't need a custom XML reader/writer these days?)

Does ET still not have a pretty output option for indentation, line breaks, etc? I've been using an little indentation function that Effbot posted somewhere a really long time ago. Or, we could use the XML writer in the UFO normalizer. That follows the normalization spec. (A spec that exists only in that code.)

> cc #27 (should the issue be salvaged for further optimizations?)

Yeah, let's keep it open for now. I see that as a general "Try to optimize some things." issue.

File: Lib/ufoLib/glifLib.py [hunk 1]
```diff
@@ -800,42 +803,30 @@ def validateLayerInfoVersion3Data(infoData):
...
+	formatVersion = tree.get("format")
+	if formatVersion is None:
+		raise GlifLibError("Unspecified format version in GLIF.")
 	try:
 		v = int(formatVersion)
 		formatVersion = v
 	except ValueError:
 		pass
+	if formatVersion not in formatVersions:
+		raise GlifLibError("Forbidden GLIF format version: %s" % formatVersion)
```
Comment (typesupply) [inline Lib/ufoLib/glifLib.py]: Could we change "forbidden" to "unknown"? In all likelihood, the only way a user would get this is if they were using an older ufoLib to read a newer GLIF. That's more of an "unknown" situation.

**Comment by typesupply** [comment]
This looks good to me.

File: Lib/ufoLib/glifLib.py (hunk 1 above)
Comment (adrientetar) [inline Lib/ufoLib/glifLib.py]: Well it's not necessarily unknown given that formatVersions is an argument that serves to forbid certain versions (mostly to forbid GLIF2 in UFO2). Sure you could have formatVersion=3 which is unknown but here is mostly about limiting GLIF versions for older versions of the standard, I reckon.

Comment (typesupply) [inline Lib/ufoLib/glifLib.py]: Fair enough.

Comment (adrientetar) [inline Lib/ufoLib/glifLib.py]: I think 'Illegal' is the word for Forbidden or Unknown.

**Pull request merged**